### PR TITLE
ElemType-only patch

### PR DESCRIPTION
**Pull request by cv3d**
DO NOT MERGE! This PR is for demonstration purposes only..

### This pullrequest changes
ElemType patch with zero changes from `CV_8U` to `CV_8UC1` etc

### CODE REVIEW

File: modules/core/include/opencv2/core/affine.hpp [hunk 1]
```diff
@@ -300,23 +300,22 @@ namespace cv
...
-               channels     = 16,
-               fmt          = traits::SafeFmt<channel_type>::fmt + ((channels - 1) << 8)
+        static const bool       generic_type = false;
+        static const int        channels     = 16;
+        static const int        fmt          = traits::SafeFmt<channel_type>::fmt + ((channels - 1) << 8);
 #ifdef OPENCV_TRAITS_ENABLE_DEPRECATED
-               ,depth        = DataType<channel_type>::depth
-               ,type         = CV_MAKETYPE(depth, channels)
+        static const ElemType   depth        = DataType<channel_type>::depth;
+        static const ElemType   type         = CV_MAKETYPE(depth, channels);
```
[the remark this answers is not in the document]
Comment (cv3d) [reply modules/core/include/opencv2/core/affine.hpp]: I think i pointed to that in some previous comment! His request boils down to the very core of the reason why `type()` and `.depth()` exists...
I hope he can realize that it is just non-sense to make depth and type the same.

**Comment by vpisarev** [comment]
thanks! the goal of this exercise is to figure out what could be the _minimal_ damage in the case the patch is accepted. So, let's reduce the patch as much as possible. Please, see the further comments to the patch

File: modules/core/src/arithm.cpp [hunk 2]
```diff
@@ -61,7 +61,7 @@ void convertAndUnrollScalar( const Mat& sc, int buftype, uchar* scbuf, size_t bl
 {
     int scn = (int)sc.total(), cn = CV_MAT_CN(buftype);
     size_t esz = CV_ELEM_SIZE(buftype);
-    getConvertFunc(sc.depth(), buftype)(sc.ptr(), 1, 0, 1, scbuf, 1, Size(std::min(cn, scn), 1), 0);
+    getConvertFunc(sc.depth(), CV_MAT_DEPTH(buftype))(sc.ptr(), 1, 0, 1, scbuf, 1, Size(std::min(cn, scn), 1), 0);
```
Comment (vpisarev) [inline modules/core/src/arithm.cpp]: since getConvertFunc() takes 2 ElemType's, let's insert 2 `CV_MAT_DEPTH()` calls into that function instead of using `CV_MAT_DEPTH()` each time we call it.

File: modules/core/src/arithm.cpp [hunk 3]
```diff
@@ -596,16 +605,21 @@ static bool ocl_arithm_op(InputArray _src1, InputArray _src2, OutputArray _dst,
 #endif
 
 static void arithm_op(InputArray _src1, InputArray _src2, OutputArray _dst,
-                      InputArray _mask, int dtype, BinaryFuncC* tab, bool muldiv=false,
+                      InputArray _mask, ElemType ddepth, BinaryFuncC* tab, bool muldiv=false,
```
Comment (vpisarev) [inline modules/core/src/arithm.cpp]: please, do not rename arguments; just change the type from int to ElemType

File: modules/core/src/arithm.cpp [hunk 4]
```diff
@@ -615,14 +629,14 @@ static void arithm_op(InputArray _src1, InputArray _src2, OutputArray _dst,
...
-        !haveMask && ((!_dst.fixedType() && (dtype < 0 || CV_MAT_DEPTH(dtype) == depth1)) ||
+        !haveMask && ((!_dst.fixedType() && (ddepth == CV_TYPE_AUTO || ddepth == depth1)) ||
                        (_dst.fixedType() && _dst.type() == type1)) &&
         (src1Scalar == src2Scalar) )
     {
         _dst.createSameSize(*psrc1, type1);
         CV_OCL_RUN(use_opencl,
             ocl_arithm_op(*psrc1, *psrc2, _dst, _mask,
-                          (!usrdata ? type1 : std::max(depth1, CV_32F)),
+                          (!usrdata ? type1 : CV_MAKE_TYPE(CV_MAX_DEPTH(depth1, CV_32F), 1)),
```
Comment (vpisarev) [inline modules/core/src/arithm.cpp]: CV_MAKE_TYPE is unnecessary here

File: modules/core/src/arithm.cpp [hunk 5]
```diff
@@ -919,26 +928,26 @@ static BinaryFuncC* getAbsDiffTab()
 }
 
 void cv::add( InputArray src1, InputArray src2, OutputArray dst,
-          InputArray mask, int dtype )
+          InputArray mask, ElemType ddepth )
```
Comment (vpisarev) [inline modules/core/src/arithm.cpp]: restore the original parameter name

File: modules/core/src/arithm.cpp [hunk 6]
```diff
@@ -1658,7 +1671,7 @@ static bool ocl_inRange( InputArray _src, InputArray _lowerb,
             BinaryFunc sccvtfunc = getConvertFunc(ldepth, CV_32S);
             sccvtfunc(lscalar.ptr(), 1, 0, 1, (uchar*)ilbuf, 1, Size(cn, 1), 0);
             sccvtfunc(uscalar.ptr(), 1, 0, 1, (uchar*)iubuf, 1, Size(cn, 1), 0);
-            int minval = cvRound(getMinVal(sdepth)), maxval = cvRound(getMaxVal(sdepth));
+                int minval = cvRound(getMinVal(sdepth)), maxval = cvRound(getMaxVal(sdepth));
```
Comment (vpisarev) [inline modules/core/src/arithm.cpp]: extra indentation

File: modules/core/src/arithm.cpp [hunk 7]
```diff
@@ -1904,7 +1918,7 @@ CV_IMPL void cvAdd( const CvArr* srcarr1, const CvArr* srcarr2, CvArr* dstarr, c
     CV_Assert( src1.size == dst.size && src1.channels() == dst.channels() );
     if( maskarr )
         mask = cv::cvarrToMat(maskarr);
-    cv::add( src1, src2, dst, mask, dst.type() );
+    cv::add( src1, src2, dst, mask, dst.depth() );
```
Comment (vpisarev) [inline modules/core/src/arithm.cpp]: restore `.type()`

File: modules/core/src/array.cpp [hunk 8]
```diff
@@ -94,7 +93,7 @@ cvSetIPLAllocators( Cv_iplCreateImageHeader createHeader,
 
 // Creates CvMat and underlying data
 CV_IMPL CvMat*
-cvCreateMat( int height, int width, int type )
+cvCreateMat(int height, int width, int type)
```
Comment (vpisarev) [inline modules/core/src/array.cpp]: unnecessary whitespace change

File: modules/core/src/convert_scale.cpp [hunk 9]
```diff
@@ -447,28 +449,30 @@ namespace cv {
 
 #ifdef HAVE_OPENCL
 
-static bool ocl_normalize( InputArray _src, InputOutputArray _dst, InputArray _mask, int dtype,
+static bool ocl_normalize( InputArray _src, InputOutputArray _dst, InputArray _mask, ElemType ddepth,
```
Comment (vpisarev) [inline modules/core/src/convert_scale.cpp]: undo `dtype` to `ddepth` rename

File: modules/core/src/cuda_gpu_mat.cpp [hunk 10]
```diff
@@ -494,18 +494,18 @@ GpuMat& cv::cuda::GpuMat::setTo(Scalar s, InputArray _mask, Stream& _stream)
...
     CV_UNUSED(_stream);
     throw_no_cuda();
 }
 
-void cv::cuda::GpuMat::convertTo(OutputArray _dst, int rtype, double alpha, double beta, Stream& _stream) const
+void cv::cuda::GpuMat::convertTo(OutputArray _dst, ElemType ddepth, double alpha, double beta, Stream& _stream) const
 {
     CV_UNUSED(_dst);
-    CV_UNUSED(rtype);
+    CV_UNUSED(ddepth);
```
Comment (vpisarev) [inline modules/core/src/cuda_gpu_mat.cpp]: undo argument rename

**Comment by cv3d** [comment]
@vpisarev Could you please reopen this PR to allow me addressing your concerns?
It seems I did not receive email updates about it for some reason or another..

**Comment by cv3d** [comment]
@vpisarev A kind reminder

**Comment by vpisarev** [comment]
@cv3d, let's get back to your PRs after OpenCV 4.0 release
